nixos/tests/tahoe.nix

```diff
@@ -0,0 +1,49 @@
+import ./make-test-python.nix ({ pkgs, ... }:
+{
+  name = "tahoe";
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ exarkun ];
+  };
+
+  nodes.machine =
+    { ... }:
+    { services.tahoe.nodes.some_storage = {
+        nickname = "some-storage";
+        tub.port = 12345;
+        storage = {
+          enable = true;
+          reservedSpace = "2G";
+          settings = {
+            "expire.enabled" = "true";
+            "expire.mode" = "age";
+            "expire.override_lease_duration" = "1day";
+          };
+        };
+      };
+      users.users."tahoe.some_storage".group = "tahoe.some_storage";
+      users.groups."tahoe.some_storage" = {};
```

Comment:
I don't know if the fact that the Tahoe-LAFS service requires this user/group configuration is in line with current NixOS policy or just a bug in the Tahoe-LAFS service.

Reply:
Usually, enabling a service should create all necessary users and groups. Setting any options external to the enabled module should not be required.

Of course, it would be even better if the application could just run with `DynamicUser = true`, as this is both simpler and more secure.

Reply:
Thanks!  On closer inspection, I found that master version of the service creates users without groups but then doesn't actually make the services run as the users it creates (they run as root).

So I made the service create the necessary groups and also run the services as the users it is creating.

I looked at using `DynamicUser` and that looks like a good future improvement.  I am not sure how to migrate existing installations to such a configuration though.  I believe using `DynamicUser` will require having the Tahoe-LAFS state beneath `/var/lib/<something>` but currently it is beneath `/var/db/tahoe-lafs`.

I suppose there could be two versions of the service, side-by-side, with `system.stateVersion` choosing between the non-DynamicUser and DynamicUser versions?

If that makes sense I might try it as a follow-up.

Reply:
stateVersion is a code smell imo. A possibility is to have a preStart script look if /var/db/foo exists and move it to /var/lib/foo if so.

Reply:
That seems fairly straightforward - but it results in an automatic irreversible state upgrade (admittedly a simple one) and isn't the point of `system.stateVersion` exactly to avoid those?

Reply:
Also, I suppose that it isn't necessarily simple.  If /var/lib and /var/db are on different filesystems then there is a non-trivial chance of the move failing.  A storage-enabled Tahoe-LAFS node should be expected to have a lot of data in its node directory (arbitrarily large amounts, basically).

Reply:
Sometimes it can be worth mentioning a breaking change in release notes and requiring a minimal manual interaction from the user, like `sudo mv oldstatedir newstatedir`. This avoids code smell, maintenance burden, etc...

This approach should be used sparingly and only when it greatly simplifies things.